ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionService.scala

```diff
@@ -0,0 +1,707 @@
+          party <- ledger.allocateParty()
+          request = ledger.getTransactionsRequest(Seq(party))
+          requestWithEmptyFilter = request.update(_.filter.filtersByParty := Map.empty)
+          failure <- ledger.flatTransactions(requestWithEmptyFilter).failed
+        } yield {
+          assertGrpcError(failure, Status.Code.INVALID_ARGUMENT, "filtersByParty cannot be empty")
+        }
+    }
+
+  private[this] val completeOnLedgerEnd = LedgerTest(
+    "TXCompleteOnLedgerEnd",
+    "A stream should complete as soon as the ledger end is hit") { ledger =>
+    val transactionsToSubmit = 14
+    for {
+      party <- ledger.allocateParty()
+      transactionsFuture = ledger.flatTransactions(party)
+      _ <- Future.sequence(Vector.fill(transactionsToSubmit)(for {
+        contract <- ledger.create(party, Dummy(party))
+      } yield contract))
+      _ <- transactionsFuture
+    } yield {
+      // doing nothing: we are just checking that `transactionsFuture` completes successfully
+    }
+  }
+
+  private[this] val processInTwoChunks = LedgerTest(
+    "TXProcessInTwoChunks",
+    "Serve the complete sequence of transactions even if processing is stopped and resumed") {
+    ledger =>
+      val transactionsToSubmit = 5
+      for {
+        party <- ledger.allocateParty()
+        _ <- Future.sequence(Vector.fill(transactionsToSubmit)(for {
+          contract <- ledger.create(party, Dummy(party))
+        } yield contract))
+        endAfterFirstSection <- ledger.currentEnd()
+        firstSectionRequest = ledger
+          .getTransactionsRequest(Seq(party))
+          .update(_.end := endAfterFirstSection)
+        firstSection <- ledger.flatTransactions(firstSectionRequest)
+        _ <- Future.sequence(Vector.fill(transactionsToSubmit)(for {
+          contract <- ledger.create(party, Dummy(party))
+        } yield contract))
+        endAfterSecondSection <- ledger.currentEnd()
+        secondSectionRequest = ledger
+          .getTransactionsRequest(Seq(party))
+          .update(_.begin := endAfterFirstSection, _.end := endAfterSecondSection)
+        secondSection <- ledger.flatTransactions(secondSectionRequest)
+        fullSequence <- ledger.flatTransactions(party)
+      } yield {
+        val concatenation = Vector.concat(firstSection, secondSection)
+        assert(
+          fullSequence == concatenation,
+          s"The result of processing items in two chunk should yield the same result as getting the overall stream of transactions in the end but there are differences. " +
+            s"Full sequence: ${fullSequence.map(_.commandId).mkString(", ")}, " +
+            s"first section: ${firstSection.map(_.commandId).mkString(", ")}, " +
+            s"second section: ${secondSection.map(_.commandId).mkString(", ")}"
+        )
+      }
+  }
+
+  private[this] val identicalAndParallel = LedgerTest(
+    "TXParallel",
+    "The same data should be served for two identical, parallel requests") { ledger =>
```

Comment:
The test runs 10 parallel requests, no?
```suggestion
    "The same data should be served for more than 1 identical, parallel requests") { ledger =>
```

Reply:
I'm :it:, I love Copy :spaghetti: .